pkg/listwatch/listinformer.go

```diff
@@ -0,0 +1,196 @@
+
+package listwatch
+
+import (
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/tools/cache"
+)
+
+type listLenListerWatcher struct {
+	listChan      chan int
+	listerWatcher cache.ListerWatcher
+}
+
+func (lllw *listLenListerWatcher) List(options metav1.ListOptions) (runtime.Object, error) {
+	lllw.listChan <- -1
+	<-lllw.listChan
+	result, err := lllw.listerWatcher.List(options)
+	if err != nil {
+		return result, err
+	}
+	lllw.listChan <- meta.LenList(result)
+	return result, err
+}
+
+func (lllw *listLenListerWatcher) Watch(options metav1.ListOptions) (watch.Interface, error) {
+	return lllw.listerWatcher.Watch(options)
+}
+
+type listerInformer struct {
+	namespace  string
+	informer   cache.SharedInformer
+	events     chan<- *watch.Event
+	stopped    chan struct{}
+	hasStopped bool
+	draining   bool
+	drained    chan<- string
+	listChan   chan int
+	listCount  int
+	addCount   int
+	lock       sync.RWMutex
+}
+
+func newListerInformer(namespace string, f func(string) cache.ListerWatcher, exampleObject runtime.Object,
+	resyncPeriod time.Duration, events chan<- *watch.Event, drained chan<- string) *listerInformer {
+
+	listCountChan := make(chan int)
+
+	informer := cache.NewSharedInformer(&listLenListerWatcher{listCountChan, f(namespace)}, exampleObject, resyncPeriod)
+
+	li := &listerInformer{
+		namespace: namespace,
+		informer:  informer,
+		events:    events,
+		stopped:   make(chan struct{}),
+		drained:   drained,
+		listChan:  listCountChan,
+		listCount: -1,
+	}
+	informer.AddEventHandler(li)
+	go func() {
+		for {
+			select {
+			case <-li.stopped:
+				return
+			case listCount := <-listCountChan:
+				func() {
+					li.lock.Lock()
+					defer li.lock.Unlock()
+					li.listCount = listCount
+					if listCount == -1 {
+						li.addCount = 0
+						listCountChan <- 0
+					}
+				}()
+			}
+		}
+	}()
+	go informer.Run(li.stopped)
+	return li
+}
+
+func (li *listerInformer) hasReachedListCount() bool {
+	li.lock.RLock()
+	defer li.lock.RUnlock()
+	return li.listCount <= li.addCount
+}
+
+func (li *listerInformer) hasSynced() bool {
+	return li.informer.HasSynced() && li.hasReachedListCount()
+}
+
+func (li *listerInformer) isDraining() bool {
+	li.lock.RLock()
+	defer li.lock.RUnlock()
+	return li.draining
+}
+
+func (li *listerInformer) isStopped() bool {
+	li.lock.RLock()
+	defer li.lock.RUnlock()
+	return li.hasStopped
+}
+
+func (li *listerInformer) drain() {
+	shouldDrain := func() bool {
+		li.lock.Lock()
+		defer li.lock.Unlock()
+		if li.draining {
+			return false
+		}
+		li.draining = true
+		li.stopInformer()
+		return true
+	}()
+	if shouldDrain {
+		go func() {
+			store := li.informer.GetStore()
+			resourcesToDrain := store.List()
+			for _, resource := range resourcesToDrain {
+				li.OnDelete(resource)
```

Comment:
I like this part, sending explicit delete events for things that fall out of the watch. might solve a few issues we encountered. we should look at implementing the same in xns-informers